### PR TITLE
tr/service_hook: prevent Update from running before Poststart finish

### LINKED ISSUE #5767: Service hook fails sometimes
### Nomad version
0.9.1

### Consul version
1.5.0

### Issue

Sometimes nomad outputs the following errors when deploying a container:

```
{"@level":"error","@message":"update hook failed","@module":"client.alloc_runner.task_runner","@timestamp":"2019-05-29T14:55:58.076399Z","alloc_id":"b5b284fb-8980-d29e-65f7-9371f3b9d15f","error":"driver doesn't support script checks","name":"consul_services","task":"my-task"}
```

```
{"@level":"error","@message":"update hook failed","@module":"client.alloc_runner.task_runner","@timestamp":"2019-05-29T14:59:43.017706Z","alloc_id":"9efb7a79-55d8-2d72-9506-96b0bf7a7050","error":"unable to get address for service \"my-service\": cannot use address_mode=\"driver\": no driver network exists","name":"consul_services","task":"my-task"}
```

In the end it works and gets registered correctly.
This happened several times with different jobs.

### Reproduction steps

I can't reproduce them all the time. It seems like a concurrency problem.

### Job file (if appropriate)

The job is a simple docker driver job with a service and a check. I saw in some issue that for address_mode driver it should be in a custom network_mode (which is the case).

This job file is the one that was used when the second error appeared (in the case of the script check error was another one). 

```
"Tasks": [
        {
          "Name": "my-task",
          "Driver": "docker",
          "User": "",
          "Config": {
            "network_mode": "my-network",
            "port_map": [
              {
                "api": 3000
              }
            ],
            "force_pull": true,
            "image": "********",
          },
          "Env": null,
          "Services": [
            {
              "Name": "${NOMAD_JOB_NAME}-${NOMAD_ALLOC_INDEX}",
              "PortLabel": "api",
              "AddressMode": "driver",
              "Tags": null,
              "CanaryTags": null,
              "Checks": [
                {
                  "Name": "${NOMAD_JOB_NAME}-${NOMAD_ALLOC_INDEX}-check-tcp-container",
                  "Type": "tcp",
                  "Command": "",
                  "Args": null,
                  "Path": "",
                  "Protocol": "",
                  "PortLabel": "",
                  "AddressMode": "",
                  "Interval": 10000000000,
                  "Timeout": 2000000000,
                  "InitialStatus": "",
                  "TLSSkipVerify": false,
                  "Method": "",
                  "Header": null,
                  "CheckRestart": {
                    "Limit": 3,
                    "Grace": 90000000000,
                    "IgnoreWarnings": false
                  },
                  "GRPCService": "",
                  "GRPCUseTLS": false
                }
              ]
            },
            {
              "Name": "${NOMAD_JOB_NAME}",
              "PortLabel": "api",
              "AddressMode": "driver",
              "Tags": null,
              "CanaryTags": null,
              "Checks": [
                {
                  "Name": "${NOMAD_TASK_NAME}-${NOMAD_ALLOC_INDEX}-check-tcp-service",
                  "Type": "tcp",
                  "Command": "",
                  "Args": null,
                  "Path": "",
                  "Protocol": "",
                  "PortLabel": "",
                  "AddressMode": "",
                  "Interval": 10000000000,
                  "Timeout": 2000000000,
                  "InitialStatus": "",
                  "TLSSkipVerify": false,
                  "Method": "",
                  "Header": null,
                  "CheckRestart": null,
                  "GRPCService": "",
                  "GRPCUseTLS": false
                }
              ]
            },
            {
              "Name": "${NOMAD_JOB_NAME}-external",
              "PortLabel": "api",
              "AddressMode": "host",
              "Tags": [],
              "CanaryTags": null,
              "Checks": [
                {
                  "Name": "${NOMAD_TASK_NAME}-${NOMAD_ALLOC_INDEX}-external",
                  "Type": "tcp",
                  "Command": "",
                  "Args": null,
                  "Path": "",
                  "Protocol": "",
                  "PortLabel": "",
                  "AddressMode": "",
                  "Interval": 10000000000,
                  "Timeout": 2000000000,
                  "InitialStatus": "",
                  "TLSSkipVerify": false,
                  "Method": "",
                  "Header": null,
                  "CheckRestart": null,
                  "GRPCService": "",
                  "GRPCUseTLS": false
                }
              ]
            }
          ],
          "Templates": [
            {
              "SourcePath": "",
              "DestPath": "secrets/file.env",
              "EmbeddedTmpl": "  ........ ",
              "ChangeMode": "restart",
              "ChangeSignal": "",
              "Splay": 5000000000,
              "Perms": "0644",
              "LeftDelim": "{{",
              "RightDelim": "}}",
              "Envvars": true,
              "VaultGrace": 15000000000
            }
          ],
          "Constraints": null,
          "Affinities": null,
          "Resources": {
            "CPU": 300,
            "MemoryMB": 300,
            "DiskMB": 0,
            "IOPS": 0,
            "Networks": [
              {
                "Device": "",
                "CIDR": "",
                "IP": "",
                "MBits": 10,
                "ReservedPorts": null,
                "DynamicPorts": [
                  {
                    "Label": "api",
                    "Value": 0
                  }
                ]
              }
            ],
            "Devices": null
          },
          "DispatchPayload": null,
          "Meta": null,
          "KillTimeout": 5000000000,
          "LogConfig": {
            "MaxFiles": 10,
            "MaxFileSizeMB": 10
          },
          "Artifacts": null,
          "Leader": false,
          "ShutdownDelay": 0,
          "KillSignal": ""
        }
      ],
```

In the case of the script check failure. The script is rendered via a template.

### Nomad Client logs (if appropriate)
```
May 29 14:59:43 w-a3d6cd97-0002 nomad: {"@level":"error","@message":"update hook failed","@module":"client.alloc_runner.task_runner","@timestamp":"2019-05-29T14:59:43.017706Z","alloc_id":"9efb7a79-55d8-2d72-9506-96b0bf7a7050","error":"unable to get address for service \"my-service\": cannot use address_mode=\"driver\": no driver network exists","name":"consul_services","task":"my-task"}
May 29 14:59:43 w-a3d6cd97-0002 nomad: 2019/05/29 14:59:43.017810 [INFO] (runner) creating new runner (dry: false, once: false)
May 29 14:59:43 w-a3d6cd97-0002 nomad: 2019/05/29 14:59:43.017969 [INFO] (runner) creating watcher
May 29 14:59:43 w-a3d6cd97-0002 nomad: 2019/05/29 14:59:43.018091 [INFO] (runner) starting
May 29 14:59:43 w-a3d6cd97-0002 nomad: 2019/05/29 14:59:43.039351 [INFO] (runner) rendered "(dynamic)" => "/opt/nomad/data/alloc/9efb7a79-55d8-2d72-9506-96b0bf7a7050/my-task/secrets/file.env"
May 29 14:59:44 w-a3d6cd97-0002 consul: 2019/05/29 14:59:44 [INFO] agent: Synced check "89eac5f576aa976e8b883d4ee1f7f0bbd5cc6bdc"
May 29 14:59:44 w-a3d6cd97-0002 nomad: {"@level":"info","@message":"created container","@module":"client.driver_mgr.docker","@timestamp":"2019-05-29T14:59:44.616571Z","container_id":"e613b13dccd874babb3950f8b1119445fde313eb908007233b7cc37c4e0b407e","driver":"docker"}
```

If you want more info I could try a get some more.

Thanks.

### PR DESCRIPTION
**Pull request by nickethier**
The service hook requires fields initialized during Poststart when executing other lifecycle functions. In particular Update can be called at any time including before Poststart. This manifests as an error because the driver network information has not been set yet.

fixes #5767

### CODE REVIEW

File: client/allocrunner/taskrunner/service_hook.go [hunk 1]
```diff
@@ -91,12 +95,17 @@ func (h *serviceHook) Poststart(ctx context.Context, req *interfaces.TaskPoststa
 	// Create task services struct with request's driver metadata
 	workloadServices := h.getWorkloadServices()
 
+	defer func() { h.initialRegistration = true }()
```
Comment (notnoop) [inline client/allocrunner/taskrunner/service_hook.go:98]: This doesn't need to be in defer, specially given we don't care if RegisterWorkflow succeeds and `h.mu` is held for the entire duration of functions call including consul call.

**Comment by nickethier** [comment]
Shout out to @42wim for doing the leg work to track down this bug :tada:

File: client/allocrunner/taskrunner/service_hook.go [hunk 2]
```diff
@@ -91,12 +95,17 @@ func (h *serviceHook) Poststart(ctx context.Context, req *interfaces.TaskPoststa
...
 	return h.consul.RegisterWorkload(workloadServices)
 }
 
 func (h *serviceHook) Update(ctx context.Context, req *interfaces.TaskUpdateRequest, _ *interfaces.TaskUpdateResponse) error {
 	h.mu.Lock()
 	defer h.mu.Unlock()
+	if !h.initialRegistration {
+		// no op since initial registration has not finished
+		return nil
+	}
```
Comment (notnoop) [inline client/allocrunner/taskrunner/service_hook.go:108]: Might make sense to update service hook fields but avoid calling out to consul.  I'm thinking of the case where a client receives an in-place update of a task before it starts - we should honor the new `TaskEnv`/`ShutdownDelay`/etc in the PostStart invocation.

File: client/allocrunner/taskrunner/service_hook_test.go [hunk 3]
```diff
@@ -1,11 +1,37 @@
...
+	"github.com/stretchr/testify/require"
 )
 
 // Statically assert the stats hook implements the expected interfaces
 var _ interfaces.TaskPoststartHook = (*serviceHook)(nil)
 var _ interfaces.TaskExitedHook = (*serviceHook)(nil)
 var _ interfaces.TaskPreKillHook = (*serviceHook)(nil)
 var _ interfaces.TaskUpdateHook = (*serviceHook)(nil)
+
+func TestUpdate_beforePoststart(t *testing.T) {
```
Comment (notnoop) [inline client/allocrunner/taskrunner/service_hook_test.go:20]: Good simple test!

File: client/allocrunner/taskrunner/service_hook.go [hunk 4]
```diff
@@ -164,6 +173,7 @@ func (h *serviceHook) Exited(context.Context, *interfaces.TaskExitedRequest, *in
 	h.mu.Lock()
 	defer h.mu.Unlock()
 	h.deregister()
+	h.initialRegistration = false
```
Comment (notnoop) [inline client/allocrunner/taskrunner/service_hook.go:176]: Also, in `Stop()` or move to `deregister`?

File: client/allocrunner/taskrunner/service_hook_test.go [hunk 5]
```diff
@@ -34,4 +34,17 @@ func TestUpdate_beforePoststart(t *testing.T) {
...
+	// so Update should again wait on Poststart.
+
+	require.NoError(t, hook.Exited(context.Background(), &interfaces.TaskExitedRequest{}, &interfaces.TaskExitedResponse{}))
+	require.Len(t, c.GetOps(), 4)
+	require.NoError(t, hook.Update(context.Background(), &interfaces.TaskUpdateRequest{Alloc: alloc}, &interfaces.TaskUpdateResponse{}))
+	require.Len(t, c.GetOps(), 4)
+	require.NoError(t, hook.Poststart(context.Background(), &interfaces.TaskPoststartRequest{}, &interfaces.TaskPoststartResponse{}))
+	require.Len(t, c.GetOps(), 5)
+	require.NoError(t, hook.Update(context.Background(), &interfaces.TaskUpdateRequest{Alloc: alloc}, &interfaces.TaskUpdateResponse{}))
+	require.Len(t, c.GetOps(), 6)
```
Comment (notnoop) [inline client/allocrunner/taskrunner/service_hook_test.go:48]: Maybe also test PreKilled then update after the last update?

**Pull request merged**